core-processor/src/processor.rs

```diff
@@ -42,33 +49,197 @@ enum SuccessfulDispatchResultKind {
+        &self.gas_counter
+    }
+}
+
+/// Defines result variants of the function `prepare`.
+#[allow(clippy::large_enum_variant)]
+pub enum PrepareResult {
+    /// Successfully precharged for memory pages.
+    Ok {
+        /// A context for `process` function.
+        context: PreparedMessageExecutionContext,
+        /// A set with numbers of memory pages which contain some data.
+        pages_with_data: BTreeSet<PageNumber>,
+    },
+    /// Required function is not exported. The program will not be executed.
+    WontExecute(Vec<JournalNote>),
+    /// Provided actor is not executable or there is not enough gas for memory pages size.
+    Error(Vec<JournalNote>),
+}
+
+/// Prepares environment for the execution of a program.
+/// Checks if there is a required export and tries to pre-charge for memory pages.
+/// Returns either `PreparedMessageExecutionContext` for `process` or an array of journal notes.
+/// See `PrepareResult` for details.
+pub fn prepare(
     block_config: &BlockConfig,
     execution_context: MessageExecutionContext,
-) -> Vec<JournalNote> {
+) -> PrepareResult {
     let MessageExecutionContext {
         actor,
         dispatch,
         origin,
         gas_allowance,
+        subsequent_execution,
     } = execution_context;
     let Actor {
         balance,
         destination_program,
         executable_data,
     } = actor;
 
-    match check_is_executable(executable_data, &dispatch) {
-        Err(exit_code) => process_non_executable(dispatch, destination_program, exit_code),
-        Ok(data) => process_executable::<A, E>(
-            origin,
-            gas_allowance,
-            data,
+    let data = match check_is_executable(executable_data, &dispatch) {
+        Err(exit_code) => {
+            return PrepareResult::Error(process_non_executable(
+                dispatch,
+                destination_program,
+                exit_code,
+            ))
+        }
+        Ok(data) => data,
+    };
+
+    let program = &data.program;
+    let program_id = program.id();
+    let mut gas_counter = GasCounter::new(dispatch.gas_limit());
+    if !program.code().exports().contains(&dispatch.kind()) {
+        return PrepareResult::WontExecute(process_success(
+            SuccessfulDispatchResultKind::Success,
+            DispatchResult::success(dispatch, program_id, gas_counter.into()),
+        ));
+    }
+
+    let mut gas_allowance_counter = GasAllowanceCounter::new(gas_allowance);
+    let memory_size = match executor::charge_gas_for_pages(
+        &block_config.allocations_config,
+        &mut gas_counter,
+        &mut gas_allowance_counter,
+        program.get_allocations(),
+        program.static_pages(),
+        dispatch.context().is_none() && matches!(dispatch.kind(), DispatchKind::Init),
+        subsequent_execution,
+    ) {
+        Ok(size) => size,
+        Err(reason) => {
+            log::debug!("failed to charge for memory pages: {:?}", reason);
```

Comment:
```suggestion
            log::debug!("Failed to charge for memory pages: {:?}", reason);
```
I'd also add logs for success charge